refactor(startup): decouple loginscreen from model elements

Kribylet opened the pull request:
LoginScreen was directly modifying Profile, and the overall startup process of qTox was confusing. This commit aims to simplify the code and reduce overall singleton dependency during startup. The scope of the commit could not be allowed to grow too large; as such there are awkward compromises in places due to the structure of singletons that could not be changed at this time. These should not be any worse than what was already in place.

Moved audio instance into Nexus instance. This was a necessary evil to concentrate related logic and facilitate a future shifting of the Nexus object responsibilities elsewhere.

Stop treating singleton instances of Nexus, Profile and Settings as singletons during main.cpp and related processes as much as possible. This does not break Singleton functionality elsewhere, but starts the process of replacing Singleton reliance with a central model object:

- If getInstance() for a singleton is used, keep and use that pointer during function call. This enables passing the object by reference easily later.

-  If getInstance() calls are used to manipulate singleton variables, prefer an emit-to-slot solution whenever possible. Due to deep nesting preventing connect() calls or heavily singleton dependent existing solution this may be non-trivial, but this should become simpler as changes propagate through the project elsewhere.

Remove code duplication related to initial LoginScreen vs. returning to LoginScreen later. This favored the Nexus solution because it is called in code elsewhere, but as above this concentration of logic should simplify moving out of the Nexus design pattern in the long term.

Made Logout action select logged out profile by default.

Profile creation and loading is now performed using signal emits rather than immediate singleton access. This means LoginScreen no longer modifies Profile instance directly.

Made AutoLogin checkbox during LoginScreen use signals instead.

Lifted makeToxPortable initialization logic into separate utility function for future decoupling of global settings vs. profile (eg. model specific) settings. This decoupling would simplify startup behavior as a Settings instance would not be required for initial file reading.

Tests performed:
 Login (default)
 Login (-l)
 Login (-p <profile>)
 Login (create new profile)
 Login (load profile)
 Login (no existing profiles)
 Login (import profile)
 Login (autologin)
 Login (autologin checkbox)
 Login (locale translation)
 Login (quit)

 MainGUI (connect)
 MainGUI (add friend)
 MainGUI (remove friend)
 MainGUI (create group)
 MainGUI (leave group)
 MainGUI (chat)
 MainGUI (separate friend chat window)
 MainGUI (logout)
 MainGUI (quit)
 MainGUI (change profile settings)
 MainGUI (change global settings)

Concerns:
 Mac OSX compatability due to moving Mac specific logic out of Nexus and changing call order. I have not been able to test those changes.

Fixes #5707

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5711)
<!-- Reviewable:end -->

Diadlo commented:
@Kribylet Thanks for your refactoring. Is it possible to split your commit on smaller? Once all changes are messed up in one commit it's **very** hard to review

Also, Travis build [failed](https://travis-ci.org/qTox/qTox/jobs/548646379#L7241) on mac os

Kribylet commented:
I am splitting the commit right now. I see what failed on the Mac build, and I will attempt to fix that to the extent that I can without a system to review it on. If Travis builds from then I think this PR would still have to be tested on Mac to be sure it works.

sudden6 commented on src/presenter/presenter.h:
```diff
@@ -0,0 +1,69 @@
+//
```
Please use the qTox copyright notice: https://github.com/qTox/qTox/blob/3fd4ce5952211d39a5c7561ff8c9144bf00f8d29/src/model/message.cpp#L1

Kribylet replied:
fixed

sudden6 commented on src/presenter/presenter.cpp:
```diff
@@ -0,0 +1,215 @@
+//
```
Please use the qTox Copyright notice: https://github.com/qTox/qTox/blob/3fd4ce5952211d39a5c7561ff8c9144bf00f8d29/src/model/message.cpp#L1

Kribylet replied:
fixed

sudden6 commented on src/presenter/presenter.h:
```diff
@@ -0,0 +1,69 @@
+
+    void setupEnvironment();
+
+public slots:
```
there are no slots here, so this is not needed

Kribylet replied:
fixed

sudden6 commented on src/widget/loginscreen.cpp:
```diff
@@ -72,45 +73,56 @@ LoginScreen::~LoginScreen()
+    // If we are in the bootstrap, returning -1 will give us something to exit with in main.cpp
+    this->setResult(-1);
+    // If we are in application exec, we can quit by closing it, instead.
+    qApp->quit();
```
I don't think it's a good idea to kill the application from here, why not use the return value for this case too?

Kribylet replied:
Because I do not believe it is possible at this time :/ 
In the first case we have called loginScreen as part of the initial startup, in which case we are about to call a->exec() (in main.cpp) and enter an empty application running indefinitely. The return value in that case will allow us to return early instead.

In the other case we have logged in (and thus are operating from the context of QApplication "a" in main.cpp) and then logged out. A closeEvent of LoginScreen is not sufficient at that stage to halt execution of the application, and nothing is listening to the return value either since showLogin was called by InvokeMethod from profileinfo.

I tried a few approaches to resolve this issue (Changing qApp->setQuitOnLastWindowClosed(false); to true, for one) but this was the only one I found that worked. I don't mean to say there couldn't be a better solution, but I could not find a better one.

Thank you for your other points, I will fix them ASAP. I also think I have a solution for the widget related mac OSX issue, again, at least to successfully build with some hands on testing recommended. Commits forthcoming, I have struggled to escape the event horizon of swedish midsummer celebrations.

sudden6 replied:
ok, this is an acceptable workaround then

sudden6 reviewed (changes requested):
Did a quick review and found some minor things, I'm leaving the architecture review to @Diadlo

Diadlo commented on src/nexus.cpp:
```diff
@@ -67,115 +58,90 @@ Nexus::~Nexus()
 
-    LoginScreen loginScreen;
-    loginScreen.exec();
+    std::unique_ptr<LoginScreen> loginScreen(new LoginScreen{profileName});
```
Why instance replaced on pointer here?

Kribylet replied:
No good reason, fixed.

Diadlo commented on src/presenter/presenter.h:
```diff
@@ -0,0 +1,69 @@
+    Q_OBJECT
+
+#ifdef Q_OS_MAC
+public:
```
Why these fields are public?

Kribylet replied:
Unknown. Reverting to old Nexus implementation of Mac OSX hooks for this PR after discussion with sudden6. If you wish me to modify this code please indicate as such and I will make a separate PR.

Diadlo replied:
Ok. I did not notice, that it was moved

Diadlo commented on src/nexus.cpp:
```diff
@@ -235,20 +235,34 @@ Profile* Nexus::getProfile()
+ * @param pass New password
  */
-void Nexus::setProfile(Profile* profile)
+void Nexus::onCreateNewProfile(QString name, const QString& pass)
```
const reference for `name`

Kribylet replied:
fixed

Diadlo commented on src/nexus.cpp:
```diff
@@ -235,20 +235,34 @@ Profile* Nexus::getProfile()
+void Nexus::onLoadProfile(QString name, const QString& pass)
+{
+    Profile* p = Profile::loadProfile(name, pass);
+    emit profileLoaded(bool(p));
```
I think this signal can be split on two: `profileLoadFailed` and `profileLoaded`. It's a bit strange to have `profileLoaded(false)` because it means that profile not loaded. Also on practice slot have two branches. One to handle success and second to handle failure. 

But feel free to leave.

Kribylet replied:
Changed as indicated.

Diadlo commented on src/nexus.cpp:
```diff
@@ -235,20 +235,34 @@ Profile* Nexus::getProfile()
+    emit currentProfileChanged(p);
+}
 
+void Nexus::onLoadProfile(QString name, const QString& pass)
```
const reference for `name`

Kribylet replied:
fixed

Diadlo commented on src/nexus.h:
```diff
@@ -44,29 +46,34 @@ class Nexus : public QObject
+    void onCreateNewProfile(QString name, const QString& pass);
+    void onLoadProfile(QString name, const QString& pass);
+    int showLogin(const QString& profileName = QString());
+    void bootStrapWithProfile(Profile* p);
```
AFAIK "bootstrap" it's one word, so there shouldn't be uppercase "S"

Diadlo commented on src/persistence/profile.cpp:
```diff
@@ -105,9 +105,8 @@ Profile::Profile(QString name, const QString& password, bool isNewProfile,
-    s.setCurrentProfile(name);
-    s.saveGlobal();
-    s.loadPersonal(name, this->passkey.get());
+    // TODO(kriby): Move/refactor core initialization to remove settings dependency
```
Why settings dependency should be removed? I suppose `Settings` should be reduced to `IProfileSettings`, but not fully removed.

Kribylet replied:
My argument is as follows; In the long term, there is nothing about a user profile that implies it should hold elements of the tox connection itself. As a direct consequence of  Profile holding Core right now, the startup sequence code complexity goes up significantly. I imagine code complexity elsewhere in the project could be affected as well, at a minimum the tight coupling makes refactoring efforts to reduce singleton dependency in the project overall harder.

In the long term, I would argue that it would healthier for qTox as a project to separate class concerns concisely and ween off use of singleton access so that we can minimize unexpected adverse effects of code changes throughout the project. I added this todo in the spirit of that goal, but I admit it may take a while to reach a point where we can safely resolve it.

Diadlo replied:
> nothing about a user profile that implies it should hold elements of the tox connection itself

Yes, but we talk about settings.

I agree with all your thesis. In particular, that Settings singleton should be replaced on Dependency Injection. But in your comment, you said that settings dependency should be completely removed

Kribylet replied:
It was mostly meant as a reminder to myself to attempt to refactor that code to the effect I described at some point in the near future. I will remove it.

Diadlo commented on src/persistence/settings.cpp:
```diff
@@ -2318,12 +2331,13 @@ bool Settings::getAutoLogin() const
 }
 
-void Settings::setAutoLogin(bool state)
+void Settings::onSetAutoLogin(bool state)
```
Why the name was changed? It's one of the settings. It has a `get/set` pair

Kribylet replied:
This was changed to update and read changes through signals rather than direct object access. Now it should emit a signal when the checkbox is clicked which prompts settings to update the value internally. When the value is updated in Settings, another signal is sent back to the LoginScreen notifying it that the autologin value has changed, which updates the visual of the checkbox. This means that LoginScreen does not have to know anything about Settings to function, which makes the code more concise and modular.

A somewhat unlikely utility of this would be if the autologin property was changed elsewhere. The previous implementation would have no way of knowing a change had taken place, while this one would send a signal indicating the update. The usefulness of this implementation is much more obvious in the case of multiple windows representing the same group chat, for example.

Diadlo replied:
Again. You misunderstood my point. I'm not against using a signal/slot mechanism to interact between classes. I talk about method name. I think `Settings` shouldn't know is its methods used as an endpoint of the connection or not. The method should be `onXXX` only if they connected inside the class itself.
I.e.
```
void MyClass::init()
{
    connect(view, &MyView::autoLoginChanged, this, &MyClass::onAutoLoginChanged);
}
```
but
```
void MyClass::init()
{
    connect(view, &MyView::autoLoginChanged, service, &MyService::setAutoLogin);
}
```

Kribylet replied:
Yes, I see what you mean now. Sorry this took so long to respond to, I didn't realize Github would hide unresolved discussions when condensing multiple. Changed as indicated.

Kribylet replied:
I renamed onCurrentProfileChanged to updateProfileData well to match this line of thinking.

Diadlo commented on src/persistence/settings.cpp:
```diff
@@ -2318,12 +2331,13 @@ bool Settings::getAutoLogin() const
 
     if (state != autoLogin) {
         autoLogin = state;
+        saveGlobal();
```
Why did this add?

Kribylet replied:
Because previously clicking the autologin checkbox called saveGlobal directly through Settings::getInstance(). The emit solution I wrote above could emit saveGlobal to preserve that behavior, but since all instances of autologin changing currently would behave the same, I opted to simply call the relevant function in its own class.

Diadlo replied:
Suppose to connect `autoLoginChanged` to `saveGlobal` and remove from here

Kribylet replied:
Absolutely. Changed as indicated.

Diadlo commented on src/widget/loginscreen.cpp:
```diff
@@ -72,45 +73,56 @@ LoginScreen::~LoginScreen()
  * @brief Resets the UI, clears all fields.
  */
-void LoginScreen::reset(QString initialProfile)
+void LoginScreen::reset(QString initialProfileName)
```
const reference

Diadlo commented on src/widget/loginscreen.cpp:
```diff
@@ -72,45 +73,56 @@ LoginScreen::~LoginScreen()
-            ui->loginUsernames->setCurrentIndex(ui->loginUsernames->count() - 1);
-        }
-    }
+    QStringList allProfileNames = Profile::getAllProfileNames(); // static utility function
```
I think this comment is redundant. Feel free to leave

Also looks like the result can be const.

Diadlo commented on src/widget/loginscreen.cpp:
```diff
@@ -266,7 +246,7 @@ void LoginScreen::retranslateUi()
 void LoginScreen::onImportProfile()
 {
-    ProfileImporter* pi = new ProfileImporter(this);
+    auto pi = new ProfileImporter(this);
```
Does anyone know, why `new` used here? Variable deleted on the end of the function. I think the instance can be used. What do you think? Or at least `unique_ptr`

Kribylet replied:
Changed and tested this, PR update will use an instance as done elsewhere in code.

Kribylet replied:
fixed

Diadlo commented on src/widget/loginscreen.h:
```diff
@@ -36,22 +36,27 @@ class LoginScreen : public QDialog
+
     void windowStateChanged(Qt::WindowStates states);
-    void closed();
+    // Connect to Settings instance
```
It can be used not only by settings. I think this comment is redundant

Kribylet replied:
Removed.

Diadlo commented on src/widget/loginscreen.h:
```diff
@@ -36,22 +36,27 @@ class LoginScreen : public QDialog
     void windowStateChanged(Qt::WindowStates states);
-    void closed();
+    // Connect to Settings instance
+    void setAutoLogin(bool state);
```
Also, I prefer the name of signals with `ed` suffix. I.e. `autoLoginChanged`,`newProfileCreated`, `profileLoaded`

Kribylet replied:
Changed as indicated

Diadlo commented on src/persistence/settings.cpp:
```diff
@@ -273,6 +263,22 @@ void Settings::loadGlobal()
+    QString localSettingsPath = qApp->applicationDirPath() + QDir::separator() + globalSettingsFile;
+    bool result;
+
+    if (QFile(localSettingsPath).exists()) {
```
I prefer early return here. Feel free to ignore.
```
if (!QFile(localSettingsPath).exists()) {
    return false;
}
```

Kribylet replied:
Changed as indicated

Diadlo replied:
The meaning of the early return is to get rid of the `else` branch and its indent. I.e.
```
if (!QFile(localSettingsPath).exists()) {
    return false;
}

QSettings ps(localSettingsPath, QSettings::IniFormat);
ps.setIniCodec("UTF-8");
ps.beginGroup("Advanced");
bool result = ps.value("makeToxPortable", false).toBool();
ps.endGroup();
return result;
```

Kribylet replied:
Fixed

Diadlo commented on src/persistence/settings.cpp:
```diff
@@ -143,9 +143,13 @@ void Settings::loadGlobal()
-        notifySound = s.value("notifySound", true).toBool(); // note: notifySound and busySound UI elements are now under UI settings
+        notifySound =
+            s.value("notifySound", true)
+                .toBool(); // note: notifySound and busySound UI elements are now under UI settings
```
Probably better to move comment on the line above but save code formatting

Kribylet replied:
Changed as indicated

Kribylet committed:
refactor(startup): decouple loginscreen from model

Kribylet committed:
refactor(startup): check parser args with empty(), not size()

Kribylet committed:
refactor(startup): display last used profile when logging out

Kribylet committed:
refactor(startup): remove unnecessary arg. passing in loadDatabase

Kribylet committed:
refactor(startup): extract startup makeToxPortable logic

This facilitates later use without instantiating Settings object during startup

Kribylet committed:
refactor(startup): condense and rename scanProfiles

Kribylet committed:
refactor(startup): handle autoLogin checkbox with signals

Kribylet committed:
refactor(startup): use instance during profile import

Kribylet committed:
refactor(startup): clang-format changes to edited files

Kribylet committed:
refactor(startup): return early in makeToxPortable logic

Diadlo commented:
A few small notes. Otherwise LGTM

sudden6 commented on src/widget/loginscreen.cpp:
```diff
@@ -245,10 +245,7 @@ void LoginScreen::retranslateUi()
-
-    if (pi->importProfile())
+    ProfileImporter pi(this);
+    if (pi.importProfile())
```
braces required

Kribylet replied:
fixed

sudden6 reviewed (commented):
Not much to add to the review of @Diadlo

sudden6 reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> from my side, @Diadlo please merge if also from your side

> Reviewed 1 of 13 files at r1, 4 of 11 files at r2, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5711#-:-LiAVR1H1Hx2RNmQtnys:bus6ktt)* status: :shipit: complete! 1 of 1 LGTMs obtained



<!-- Sent from Reviewable.io -->
